libs/sql-parser/src/main/antlr/SqlBaseParser.g4

```diff
@@ -155,12 +155,19 @@ queryNoWith
     ;
 
 limitClause
-    : LIMIT (limit=parameterOrInteger | ALL)
-    | FETCH (FIRST | NEXT) (limit=parameterOrInteger) (ROW | ROWS) ONLY
+    : LIMIT (limit=offsetOrLimitParameter | ALL)
+    | FETCH (FIRST | NEXT) (limit=offsetOrLimitParameter) (ROW | ROWS) ONLY
     ;
 
 offsetClause
-    : OFFSET offset=parameterOrInteger (ROW | ROWS)?
+    : OFFSET offset=offsetOrLimitParameter (ROW | ROWS)?
+    ;
+
+offsetOrLimitParameter
+    : parameterOrInteger                                                             #offsetOrLimitParam
+    | parameterOrLiteral CAST_OPERATOR dataType                                      #offsetOrLimitDoubleColonCast
+    | CAST OPEN_ROUND_BRACKET expr AS dataType CLOSE_ROUND_BRACKET                   #offsetOrLimitCast
+    | TRY_CAST OPEN_ROUND_BRACKET expr AS dataType CLOSE_ROUND_BRACKET               #offsetOrLimitCast
```

Comment:
What do you think of adding cast support in general to `parameterOrInteger`?

Maybe we could even special case the cast in the `AstBuilder` to only support a limited set of casts (only numbers).

Adding more definitions makes it harder to reason about for users (and us) where/what is supported and leads to inconsistency. Looking at where `parameterOrInteger` is used I don't think it would harm to support casts in those places too.

Reply:
I checked this as well, and though it might be too much to add cast support to parameter values, but on the other hand, why not, we add this support and simplify the grammar, I'll rework this.